### PR TITLE
Serve pipeline configuration from server

### LINKED ISSUE #454: Implement more dynamic pipeline editor configuration
Currently, the pipeline editor has static JSON configuration files. We should enable
more dynamic configuration files via REST APIs. 

Files today:
https://github.com/elyra-ai/elyra/blob/master/packages/pipeline-editor/src/palette.json
https://github.com/elyra-ai/elyra/blob/master/packages/pipeline-editor/src/properties.json

### PR DESCRIPTION
**Pull request by lresende**
This is a bring-up of #650 without any front-end updates and updates of the json files. 

Fixes #454 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/pipeline/handlers.py
Comment (kevin-bates) [inline elyra/pipeline/handlers.py:111]: Should raise HTTPError with a 400 status here.

File: elyra/pipeline/handlers.py
Comment (kevin-bates) [inline elyra/pipeline/handlers.py:112]: I think we want to call `self.finish(msg_json)` here.  Should probably set the content-type header as well.

Comment (lresende) [reply elyra/pipeline/handlers.py:112]: `finish` already sets the content-type, but explicitly doing that to follow the same pattern used in other elyra handlers

**Comment by lresende** [comment]
@marthacryan is this easy to integrate with your pr for the new pipeline-editor component?

**Review by kevin-bates — CHANGES_REQUESTED** [review]
The openAPI doc needs to be updated and tests should be implemented.

**Comment by lresende** [comment]
Do we have a simple way to test handlers by bootstrapping a server that is exposing these handlers? Could you send a reference link, please?

**Comment by lresende** [comment]
> The openAPI doc needs to be updated

Once it is documented, does it become an official/supported API? If so, I would prefer not to do so as this will evolve to a completely different API once we add the notion of "component registry", and only document the final API.

**Comment by kevin-bates** [comment]
> Once it is documented, does it become an official/supported API?

Only in the sense that everything else is "official".   (Elyra should really be using version indicators in its endpoints actually.)

> this will evolve to a completely different API once we add the notion of "component registry"

The "component registry" will change how envs are extracted from notebooks and scripts?  Interesting.

**Comment by kevin-bates** [comment]
> Do we have a simple way to test handlers by bootstrapping a server that is exposing these handlers? 

Not that I'm aware of.  I currently use the [RESTClient extension](http://restclient.net/) on Firefox if no tests exist. 

> Could you send a reference link, please?

The metadata service includes handler testing: https://github.com/elyra-ai/elyra/blob/master/elyra/metadata/tests/test_handlers.py

@kiersten-stokes and I discussed adding this in a separate PR since there was going to be a release today.  If that isn't the case, it seems like this could be included here as well.

**Comment by kevin-bates** [comment]
I'm sorry, but my last two comments are really regarding #1460 - although I suppose they do have some application here - since both tests and open API updates do not exist.

**Comment by lresende** [comment]
@kevin-bates basic tests have been added

File: elyra/pipeline/handlers.py
Comment (kevin-bates) [inline elyra/pipeline/handlers.py:106]: Please remove or convert to an appropriate log statement.

File: elyra/pipeline/handlers.py
Comment (kevin-bates) [inline elyra/pipeline/handlers.py:107]: Python doesn't really use parenthesis by default.  I suspect some linters/IDEs will flag these as redundant.

File: elyra/pipeline/resources/palette.json
Comment (kevin-bates) [inline elyra/pipeline/resources/palette.json:8]: After seeing the demo for Pipelines 2.0 in today's community meeting, should the icons for each of the palette entries be represented here?

File: elyra/pipeline/resources/properties.json
Comment (kevin-bates) [inline elyra/pipeline/resources/properties.json:164]: This doesn't look right.  Perhaps a name like `outputs.csv` or `results.csv` would be more appropriate?

File: elyra/pipeline/tests/test_handlers.py
Comment (kevin-bates) [inline elyra/pipeline/tests/test_handlers.py:34]: We should move this definition (as well as the one in the [metadata tests](https://github.com/elyra-ai/elyra/blob/7f7a1cf006232f27a6f41f252d97113d7b8b39dd/elyra/metadata/tests/conftest.py#L89-L98)) to the [parent conftest.py](https://github.com/elyra-ai/elyra/blob/master/conftest.py) since that will become the location for fixture definitions that span all subsystems.  The contents stuff will need that as well.

Note: I'm not certain if this location should be more in the elyra location, rather than the root of the repo. Will have to see how CI goes, etc.

File: elyra/pipeline/tests/test_handlers.py
Comment (kevin-bates) [inline elyra/pipeline/tests/test_handlers.py:75]: I think a strict equality test will be difficult (and see there are not calls that actually validate the message text).  I would suggest using `expected_message in str(e)` instead.  Then the dev only needs to provide the applicable segment of the message while allowing for variations (like temp filenames for example) introduced by the test framework.

File: elyra/pipeline/tests/test_handlers.py
Comment (kevin-bates) [inline elyra/pipeline/tests/test_handlers.py:86]: Same as previous (use `in` clause).

File: elyra/__init__.py
Comment (kevin-bates) [inline elyra/__init__.py:53]: This endpoint should be added to the [elyra.yaml](https://github.com/elyra-ai/elyra/blob/master/elyra/api/elyra.yaml) open api doc.

**Review by kevin-bates — COMMENTED** [review]
Good stuff.  Just had a few (mostly minor) comments.  Thanks.

File: elyra/pipeline/resources/palette.json
Comment (lresende) [reply elyra/pipeline/resources/palette.json:8]: this is a copy of what is in master today, when we merge the new pipeline 2.0 there is going to be needed another integration pr, which can handle any additional requirements that comes from that pr.

File: elyra/pipeline/tests/test_handlers.py
Comment (lresende) [reply elyra/pipeline/tests/test_handlers.py:75]: so, this utility function is a copy from metadata, which is a copy from jupyter_server. Can we get rid of these local copies and use the one from jupyter_server? should we expose those from there?

File: elyra/__init__.py
Comment (lresende) [reply elyra/__init__.py:53]: As I mentioned, this part is the one I would like to hold off until the "component registry" is done.

File: elyra/pipeline/tests/test_handlers.py
Comment (lresende) [reply elyra/pipeline/tests/test_handlers.py:86]: Same as above :)

File: elyra/pipeline/handlers.py
Comment (lresende) [reply elyra/pipeline/handlers.py:106]: My bad, left over of troubleshooting cis

File: elyra/pipeline/tests/test_handlers.py
Comment (lresende) [reply elyra/pipeline/tests/test_handlers.py:34]: had the same thought, but not sure where the location should be.

Comment (kevin-bates) [reply elyra/pipeline/tests/test_handlers.py:34]: Yeah.  I don't know pytest well enough to know if added to the root of the repo, then would all tests need to run from there or not.  Also, we should probably decide if we want users to be able to run all tests after installation - I think its a good idea - at least server unit tests.  It might make more sense to move the top-most file into elyra.

Regardless, we should remove the duplication with this PR, whether that means moving the top-most file down or creating another in elyra.  The contents/properties handler tests will need this as well.

File: elyra/pipeline/tests/test_handlers.py
Comment (kevin-bates) [reply elyra/pipeline/tests/test_handlers.py:75]: Yeah, this rings a bell now.   This function was used directly in `metadata/tests/test_utils.py` because, at that time, the server tests were external to the package.  However, since then, they've been moved back into the package so that users could run tests after `pip install`.  So, I agree, just use the one from server now that it's available.  Do you mind removing the function from the metadata/tests/test_utils and updating the import statement in metadata/tests/test_handlers?

I see none of the metadata tests use the message parameter, but there are some jupyter_server tests that use specific substitutions for the message parameter to allow equality to work - which I'm sure requires debugging to get that right, but that's fine (especially since none of our tests use it).

Comment (lresende) [reply elyra/pipeline/tests/test_handlers.py]: removed the duplication and we are now importing it from jupyter-server

File: elyra/pipeline/tests/test_handlers.py
Comment (lresende) [reply elyra/pipeline/tests/test_handlers.py:34]: Created #1578 for agreement on how to best refactor backend tests aiming for better code reuse.

**Comment by lresende** [comment]
@kevin-bates We are thinking of merging the new pipeline editor soon. Any other changes on this one before we can merge so we can start working on integrating this to the new front-end?

**Pull request merged**
